Cleanup, Refactoring, and Bug Fixes for Google Sync

ApatheticCosmos opened the pull request:
A continuation of #6146 
Refactored and broke up some functions. Fixed some bugs.
Followed review suggestions from @gymad
This work will be included in 7.11

<!--- Provide a general summary of your changes in the Title above -->

## Description
Lot's of code cleanup. Fixed a major Locale bug that made it impossible to sync outside of mm/dd/yyyy date formats. Code is much more robust now.

~~This isn't finished yet, but I wanted to get this PR out so we can start collaborating on it.~~

<!--- Describe your changes in detail -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here unless your commit contains the issue number -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How To Test This
<!--- Please describe in detail how to test your changes. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
<!--- Go over all the following points and check all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

<!--- Your pull request will be tested via Travis CI to automatically indicate that your changes do not prevent compilation. --->

<!--- If it reports back that there are problems, you can log into the Travis system and check the log report for your pull request to see what the problem was. --->

ApatheticCosmos commented:
Wow... I actually made the codacy bot happy :laughing:

horus68 commented:
Just a note on language string as it looks like it do not parse the single quotes in a value. Using the html entity in `Don&#39;t` isn't a solution either because translators will make mistakes!

ApatheticCosmos commented:
@horus68 I'm not sure what to do at this point. The language file has other examples of both escaping a single quote/apostrophe (like LBL_USE_REAL_NAMES_DESC), and using HTML codes for special symbols (like LBL_SETTINGS_URL_DESC).
I don't know of any other solutions to the problem. :frowning_face:
How should this be handled? I could spell it out as 'do not' but that can't be a solution all the time. There's too many contractions commonly used in english to ban them all.

There's one possible solution I just thought of... could we use a unicode [’] instead? It seems to work.
It looks like this in the code:
![image](https://user-images.githubusercontent.com/16560296/49162784-a4c5b900-f2f9-11e8-91af-4451ac46add7.png)

And like this when rendered by chrome:
![image](https://user-images.githubusercontent.com/16560296/49162761-937cac80-f2f9-11e8-9590-2f152745c3f3.png)

What are your thoughts on this?

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1109,25 +1261,31 @@ public function doSync($id)
+            if (!$user) {
+                throw new Exception('Unable to get User bean.');
+            }
             $user->retrieve($row['id']);
```
you can retrive the User with BeanFactory::getBean() above in one step. OR handle the case when there is no user with this ID.

ApatheticCosmos replied:
Done. One step

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -495,107 +657,67 @@ public function setGService()
-                return "skip";
-            }
+            // For two events, pass to the helper function that sorts them and returns an array of objects
+            $events_array = $this->sortEventObjects($event_1, $event_2);
```
sortEventObjects() return value could be a boolean but used as an array as in_array() parameter few line above. Please, handle the boolean return value.

ApatheticCosmos replied:
Done. Throws exception if it returns false. Should never happen.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -495,107 +657,67 @@ public function setGService()
      * @param \Google_Service_Calendar_Event $event_2 (optional) Google Event
      *
-     * @return string push(_delete), pull(_delete), skip
+     * @return string|bool push(_delete), pull(_delete), skip
```
Use typehinting at function parameters, it makes sortEventObjects() unnecessary and more simple code.

ApatheticCosmos replied:
I'm going to be working on the pushPullSkip method. Give me day or two for testing.

ApatheticCosmos replied:
The next commit will have a refactored pushPullSkip method.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -495,107 +657,67 @@ public function setGService()
-     * @return string push(_delete), pull(_delete), skip
+     * @return string|bool push(_delete), pull(_delete), skip
      */
     public function pushPullSkip($event_1, $event_2 = null)
```
method complexity is too large

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
                 case "push":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Meeting: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Meeting: ' . $meeting->name);
                     $this->pushEvent($meeting, $gevent);
```
pushEvent() returns false on failure. Please handle the error.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1079,26 +1231,26 @@ public function doSync($id)
                 case "push":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Event: ' . $gevent->getSummary());
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Event: ' . $gevent->getSummary());
                     $this->pushEvent($meeting, $gevent);
```
pushEvent() could returns false on failure. Please handle the error.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -972,9 +1127,8 @@ public function setTimezone($timezone)
     {
-
         if (isset($gEventId)) {
             $event_local->gsync_id = $gEventId;
```
What if $gEventId is null? Please handle it as it could be `null` also.

ApatheticCosmos replied:
It's allowed to be null. If it is then the record is saved with a synctime, but not a gsync_id.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
                 case "pull_delete":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Meeting: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Meeting: ' . $meeting->name);
                     $this->delMeeting($meeting);
```
delMeeting could return false as failure. Please handle the error.

ApatheticCosmos replied:
return is now passed to caller

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1079,26 +1231,26 @@ public function doSync($id)
                 case "pull_delete":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Meeting: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Meeting: ' . $meeting->name);
                     $this->delMeeting($meeting);
```
delMeeting() couls returns false on failure. Please handle the error.

ApatheticCosmos replied:
return is now passed to caller

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1079,26 +1231,26 @@ public function doSync($id)
                 case "push_delete":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Event: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Event: ' . $meeting->name);
                     $this->delEvent($gevent, $meeting->id);
```
delEvent() couls returns false on failure. Please handle the error.

ApatheticCosmos replied:
return is now passed to caller

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
                 case "push_delete":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Event: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Deleting Event: ' . $meeting->name);
                     $this->delEvent($gevent, $meeting->id);
```
delEvent could return false as failure. Please handle the error.

ApatheticCosmos replied:
return is now passed to caller

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -740,19 +875,56 @@ public function delEvent($event, $meeting_id)
 
             // This removes the gsync_id reference from the table.
-            $sql = "UPDATE Meetings SET gsync_id = '' WHERE id = '" . $meeting_id . "'";
+            $sql = "UPDATE meetings SET gsync_id = '' WHERE id = '" . $meeting_id . "'";
```
Possible sql inject here: $meeting_id concatenated into sql query how it given as function parameter. Variable should be escapend and validated.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -740,19 +875,56 @@ public function delEvent($event, $meeting_id)
+        }
+
+        // Mark all reminders where both popup and email are disabled as deleted.
+        $sql = sprintf("UPDATE reminders SET deleted = '1' WHERE popup = '0' AND email = '0' AND related_event_module_id = '%s' AND deleted = '0'", $eventIdQuoted);
```
Can you please resolve this query and previous one above (https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R912) in one step.

ApatheticCosmos replied:
I don't know that it's possible. The first query sets all popups to 0 for the reminders. The next one sets reminders with both popups and emails set to zero as deleted. It's likely that the reminder has an email.

ApatheticCosmos replied:
The next commit will have a refactored pushPullSkip method.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -678,11 +788,36 @@ public function pullEvent($event_remote, $event_local = null)
+        if (!isset($event_local) || empty($event_local)) {
+            $event = $this->createSuitecrmMeetingEvent($event_remote);
+        } else {
+            $event = $this->updateSuitecrmMeetingEvent($event_local, $event_remote);
```
$event could be boolean false if error occured in updateSuitecrmMeetingEvent(). Please handle the error.

ApatheticCosmos replied:
Yep... that means something horrible happened. Now logging it and returning false from the function, which gets handled downstream.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -678,11 +788,36 @@ public function pullEvent($event_remote, $event_local = null)
+    public function pullEvent(Google_Service_Calendar_Event $event_remote, Meeting $event_local = null)
+    {
+        if (!isset($event_local) || empty($event_local)) {
+            $event = $this->createSuitecrmMeetingEvent($event_remote);
```
$event could be boolean false if error occured in createSuitecrmMeetingEvent() by calling updateSuitecrmMeetingEvent(). Please handle the error.

ApatheticCosmos replied:
fixed same as above

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -619,9 +741,8 @@ public function pushPullSkip($event_1, $event_2 = null)
     {
-
         if (!isset($event_remote) || empty($event_remote)) {
             $event = $this->createGoogleCalendarEvent($event_local);
```
here and above (https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R750) returned value could be false as failure. Please handle the error.

ApatheticCosmos replied:
Fixed same as above

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -634,39 +755,28 @@ public function pushEvent($event_local, $event_remote = null)
         $this->setLastSync($event_local, $return->getId());
```
setLastSync() return value could be false as failure. Please handle the error.

ApatheticCosmos replied:
Done

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -790,18 +961,9 @@ public function updateSuitecrmMeetingEvent($event_local, $event_remote)
-        $res = $this->db->query($sql);
+        // Disable all popup reminders for the SuiteCRM meeting. We add them back from Google event below.
+        $event_id = $event_local->id;
+        $this->clearPopups($event_id);
```
clearPopups could return false as failure. Please handle the error.

ApatheticCosmos replied:
handled

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -763,11 +935,10 @@ public function delEvent($event, $meeting_id)
      *
      * @return Meeting SuiteCRM Meeting Bean
```
function could pass false as failure. Please update the phpdoc

ApatheticCosmos replied:
Done

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -834,10 +1002,13 @@ public function updateSuitecrmMeetingEvent($event_local, $event_remote)
+        }
         $meeting->id = create_guid();
         $meeting->new_with_id = true;
         $event_local = $this->updateSuitecrmMeetingEvent($meeting, $event_remote);
```
updateSuitecrmMeetingEvent() could return false as failure and this method returns it. Please update the phpdoc.

ApatheticCosmos replied:
Done

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -810,21 +972,27 @@ public function updateSuitecrmMeetingEvent($event_local, $event_remote)
             }
         }
-        return  $event_local;
+        return $event_local;
```
nothing guarantees the $event_local given valid SuiteCRM meeting values from $event_remote getter methods. Please validate it before use it, save it or pass it as return value.

ApatheticCosmos replied:
Now checking for valid strings and start/end times.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -874,30 +1044,15 @@ public function updateGoogleCalendarEvent($event_local, $event_remote)
-            " AND reminders.related_event_module_id = '$event_local_id'" .
-            " AND popup = '1'"
+                "", "reminders.related_event_module = 'Meetings'" .
+                " AND reminders.related_event_module_id = '$event_local_id'" .
```
possible sql injection: $event_local_id should be escaped and validated before concatenated into sql query.

ApatheticCosmos replied:
Done

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -923,7 +1078,7 @@ public function updateGoogleCalendarEvent($event_local, $event_remote)
      * @return \Google_Service_Calendar_Event Google_Service_Calendar_Event Object
      */
-    public function createGoogleCalendarEvent($event_local)
+    public function createGoogleCalendarEvent(Meeting $event_local)
```
line https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R1094: `unset($event_remote_empty); // We're done with this` is unnecessary

ApatheticCosmos replied:
Fixed.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -972,9 +1127,8 @@ public function setTimezone($timezone)
      * @return bool Success/Failure
      */
-    protected function setLastSync($event_local, $gEventId = null)
+    protected function setLastSync(Meeting $event_local, $gEventId = null)
```
in Line (https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R1137): $return should be a valid SuiteCRM guid but save() method could be return with a failure. Please validate $return variable if it's a valid ID and handle accidental errors before set the accept status as 'accept'.

ApatheticCosmos replied:
Done

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1143,22 +1301,25 @@ public function syncAllUsers()
 
         // First we populate the array of syncable users
-        $this->setSyncUsers();
+        if (!$this->setSyncUsers()) {
```
setSyncUser() can throws an exception if retriving a User is implossible. Please catch/handle the error.

ApatheticCosmos replied:
Done

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1143,22 +1301,25 @@ public function syncAllUsers()
     {
```
This is what's called by the schedule.

gymad replied:
Ok :+1:

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -146,7 +175,7 @@ public function addUser($id, $name)
     public function delUser($id)
```
I wrote that method very early on as the mirror method to addUser. I can't think of any current use for it. Should I get rid of it?

gymad replied:
Are any chance if it will be needed? If it works as expected and tested and also you think it would be usefull in a possible future request later so it's nessessary, you can keep it, but I don't see any reason. Next time please don't implement unnecessary fetures, just think about what if it has a bug somewhere what we can not realise just because there is no any use case and we can not try it before the end-users or custom developers does.. (but it's only my view, it's up to you :) )

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
         }
 
         $meetings = $this->getUserMeetings();
         if (!isset($meetings)) {
```
It can be an array with at least one meeting in it, false, or null (on error). I've changed it to 'if (empty($meetings))'

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -182,23 +211,40 @@ public function getClient($id)
+            }
         }
 
         $this->workingUser->retrieve($id);
```
I went the simplify route and had it pulled with the BeanFactory call.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -384,14 +472,13 @@ public function getGoogleEventById($event_id)
-         // If we didn't get passed an event, return null
-            return;
+            // If we didn't get passed an event, return null
+            throw new \InvalidArgumentException('event ID is empty');
```
This can be null. Can't throw an exception here.

gymad replied:
If it is the case please think about it: What does it means if this methos returns `null`? I mean this method also returns `null` when did not find any google event and also returns null if given parameter $event_id was `null`. So how the caller can understend what is going on? If a caller can not decide what was the reason of failure, then calles has not any chance to handle the error properly. At least add a log note to give a chence to the end-user to figurate out what just happend here..

gymad replied:
In other way, the problem here is: This method getGoogleEventById() called at doSync() and passing the return (which could be `null`) to pushPullSkip() method. this method try to resolve the gevent, but it returns a `false` as it should, so the $dowhat could be `false` at this point:
```
            $gevent = $this->getGoogleEventById($meeting->gsync_id);
            $dowhat = $this->pushPullSkip($meeting, $gevent);
```
and later in a dafeult case of the following switch, it will be concatenated into a log not:
```
$this->logger->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'pushPullSkip() returned unknown value: ' . $dowhat . ' for event: ' . $meeting->name);
```
but the `false` value converted to an empty string, which is very confusing to a human reader just like the end-users or custom developers who will calling this method.

ApatheticCosmos replied:
I'll find a better way to handle this. I just had to set it back to returning null because it was breaking and I was working on another part of the code.
Thank you for the explanations.

ApatheticCosmos replied:
Ok. This is how I'm handling this. Let me know what you think.

First, I'm checking in doSync if $meeting->gsync_id is empty. If it is, I'm not calling getGoogleEventById at all and declaring $gevent as null then skipping right to pushPullSkip.

Second, I'm changing getGoogleEventById so that it only ever returns the google event or null. Never false. If the google calendar service isn't set up, it throws an exception. It should NEVER get there.

I've also split out the large case statements from doSync into doAction, and reduced it to a single statement.

gymad replied:
Thank you, it sounds right, will see.
Also I wrote a small gist example to explan why and how to use exception to solve this kind of problems:
https://gist.github.com/gymad/e8ed05aaf6e7e9c05472c99d1007c5a3

ApatheticCosmos replied:
Just in time. I'm working on other parts that could use it. Thanks!

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -281,39 +323,46 @@ public function setUsersGoogleCalendar()
+     *
+     * @param Google_Service_Calendar_CalendarList $calendarList
+     */
+    protected function findSuiteCRMCalendar($calendarList)
```
Fixed.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -209,13 +255,14 @@ public function getClient($id)
-            $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Refreshing Access Token');
+            $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Refreshing Access Token');
             $refreshToken = $client->getRefreshToken();
             $client->fetchAccessTokenWithRefreshToken($refreshToken);
```
@gymad Please take a look at the recent changes I've made here. Does it look right?

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -345,8 +424,7 @@ public function getGoogleEventByMeetingId($meeting_id)
         // Make sure the calendar service is set up
-        if (!isset($this->gService)) {
-            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'The Google Service is not set up. See setGService Method.');
+        if (!$this->isServiceExists()) {
```
Currently unused. I think I'll stash it somewhere else for now.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1143,22 +1301,25 @@ public function syncAllUsers()
         if (isset($this->users) && !empty($this->users)) {
-            foreach ($this->users as $key => $value) {
+            foreach (array_keys($this->users) as $key) {
                 $return = $this->doSync($key);
```
Done.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -247,31 +295,25 @@ public function getUserMeetings()
         // if the SuiteCRM calendar doesn't exist... Create it!
-        if (!isset($this->calendarId)) {
-            $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Unable to find the SuiteCRM Google Calendar, Creating it!');
+        if (!$this->isCalendarExists()) {
```
I think I cleaned this up sufficiently. findSuiteCRMCalendar now returns the id.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
      */
     public function doSync($id)
```
Done.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1079,26 +1231,26 @@ public function doSync($id)
             switch ($dowhat) {
                 case "push":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Event: ' . $gevent->getSummary());
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Event: ' . $gevent->getSummary());
```
I think what you're referring to got fixed when I split doSync out and created doAction.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -403,44 +490,12 @@ public function getGoogleEventById($event_id)
      * @param string $event_id The Google Event ID
      *
-     * @return \Meeting SuiteCRM Meeting Bean
+     * @return \Meeting|bool SuiteCRM Meeting Bean if found, false on error, null if not found
```
Quoted

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1002,71 +1156,69 @@ protected function setLastSync($event_local, $gEventId = null)
             switch ($dowhat) {
                 case "push":
-                    $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Meeting: ' . $meeting->name);
+                    $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Pushing Meeting: ' . $meeting->name);
```
pushPullSkip() no longer can return "unknown"
This should be cleaner now.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1,4 +1,5 @@
 <?php
```
Test Coverage is only 37%, needs more test.

ApatheticCosmos replied:
Yep. I'm going to try to write some more tests. What's going to trip me up is writing tests for methods that require communication with google, and a valid sync config set up.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -972,27 +1042,75 @@ public function setTimezone($timezone)
+     * 
+     * @return bool Success/Failure
+     */
+    protected function doAction(Meeting $meeting = null, Google_Service_Calendar_Event $event = null, $action)
```
Wrong order of arguments. Last argument can not be requered if the previous arguments are optional. (also the function complexity is too large.)

ApatheticCosmos replied:
re-ordered arguments. I'll work on the complexity. The mess at the top is only used for logging. :|

ApatheticCosmos replied:
Fixed

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -209,13 +241,24 @@ public function getClient($id)
-            $client->fetchAccessTokenWithRefreshToken($refreshToken);
+            if (empty($refreshToken)) {
+                throw new Exception('Refresh token is missing');
+                return false;
```
Unreachable statement. - you can not return at the next line of throw

ApatheticCosmos replied:
Of course... that's a derp on my part. Fixing.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -209,13 +241,24 @@ public function getClient($id)
+                throw new Exception('Refresh token is missing');
+                return false;
+            } else {
+                try {
```
Too many nested block. - After a thow or return, the `else` became unnecessary.

ApatheticCosmos replied:
Fixed

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -740,48 +780,92 @@ public function delEvent($event, $meeting_id)
+
+        // Disable all popup reminders for the SuiteCRM meeting
+        $eventIdQuoted = $this->db->quoted($event_id);
+        $sql = sprintf("UPDATE reminders SET popup = '0' WHERE related_event_module_id = %s AND deleted = '0'", $eventIdQuoted);
```
Performance: these two sql query could be solved in one step, using only one query and sql request.
see:
https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R817
and:
https://github.com/salesagility/SuiteCRM/pull/6539/files#diff-034e880ea433e2cc2571088f7cbc6620R825

ApatheticCosmos replied:
Got it.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -182,23 +195,42 @@ public function getClient($id)
-            $this->workingUser = new User();
+            $this->workingUser = BeanFactory::getBean('Users', $id);
+            if (!$this->workingUser) {
+                throw new Exception('Unable to retrieve a User bean');
```
Please add a `@throws Exception` to the phpdoc above the function. 
(Also if you add a unique const code to the exception, the caller can handle the errors more easy depends on what error happend exactly. For e.g:)
```
throw new Exception('Unable to retrieve a User bean', GoogleSync::GSYNC_ERR_NO_USER);
...
...
catch (Exception $e) {
  switch ($e->getCode()) {
    case GoogleSync::GSYNC_ERR_NO_USER:
      echo 'Dear user, I am theribly sorry, but I couldn\'t resolve the requested user for you, please see the log for more info..';
      log('Google Sync error: ' . $e->getMessage());
      break;
    default:
      echo 'Dear user, something went wrong...';
      log('An error: ' . $e->getMessage());
      break;
}
```

(Also please update the phpdoc everywhere where exceptions uses so that third party developers can read your code more easy)

ApatheticCosmos replied:
I think I did one better than a const code, but let me know what you think.
I extended the Exception class for each type of exception (11 in all). This means the code is baked into the exception, and we can use multiple catch statements to see what exceptions were thrown.
I also added a test for each new exception.

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -740,48 +780,92 @@ public function delEvent($event, $meeting_id)
 
             $this->syncedList[] = $meeting_id;
             return true;
         } else {
```
if you `return` from the function above one line of an `else`, the `else` beacame nonsence and unnecessary.
(Also, please use less `return`s /per methods so that the third party developers can debug your code more easy, it also makes less complexity of your code. - Not only here, everywhere..)

gymad commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1143,22 +1231,36 @@ public function syncAllUsers()
+            $this->logger->warn(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . $failures . ' failure(s) found at syncAllUsers method.');
             return false;
         } else {
             return true;
```
This file contains a single class in 1266 lines, which is too large. Try to split out this class into small units or classes.

ApatheticCosmos replied:
Split into GoogleSyncBase and GoogleSync.

gymad commented on modules/Configurator/views/view.edit.php:
```diff
@@ -1,7 +1,4 @@
 <?php
```
This file is not coveraged. (0%) It could be better, please add test for these parts where you having changes or additional code for this file at least.

ApatheticCosmos replied:
It seems the Configurator class has no tests whatsoever. I also cannot find any tests that specifically test functions in view.*.php files. 
Soooo... I'm creating a different pull request with tests for Configurator. It'll have at least a few methods tested so it's not completely useless. See pull #6701
I think I'll need some help testing some of the more complicated methods, as well has how to test the view.edit.php file.

gymad commented on modules/Users/GoogleApiKeySaverEntryPoint.php:
```diff
@@ -0,0 +1,259 @@
+<?php
```
This file having 43% coverage. It could be better. Please, add unit test to cover the rest of the new codes.

ApatheticCosmos replied:
@gymad 
There were some tests you wrote before the switch from codecept to phpunit. They're still accessible but were never moved over. I'm converting them now. I'll see what more I can build onto those.

ApatheticCosmos replied:
@gymad 
With my latest commit:
GoogleApiKeySaverEntryPoint  Lines:  75.90% ( 63/ 83)

ApatheticCosmos commented:
@gymad 
I've done some work on this. The biggest thing I did was move almost all the methods from public to protected. The only methods that are public are doSync() and syncAllUsers(). Hopefully this reduces the exposure. I changed the tests to use reflection as well.
I also cleaned up a couple of the review items.

ApatheticCosmos commented:
I wish codacy-bot wasn't so slow to run.
I'm using phpmd locally which is scoring it like this:
![image](https://user-images.githubusercontent.com/16560296/49894638-6e6d5b00-fe1c-11e8-934f-231479b31314.png)

Mausino commented:
Maybe will fine if travis will also support php 7.3 http://php.net/supported-versions.php which Initial Release was in 6.12.2018

gymad commented on tests/unit/include/GoogleSync/GoogleSyncTest.php:
```diff
@@ -8,7 +8,7 @@ class GoogleSyncTest extends \SuiteCRM\StateCheckerUnitAbstract
 
     /** @var ReflectionClass */
-    protected static $relection;
+    protected static $reflection;
```
I like your reflection class solution to access protected and private things. :+1:

gymad commented:
Hi @ApatheticCosmos 
Code looks much better, only a test failing at this moment. Don't take too much focus to codacy, we will ignoring the false positive issues like Worpress issues, long/short variable names etc. more important codacy issues are: complecity and unused variables, superglobals access etc. (Codacy issues also doesn't high priority if it's coming up in tests folder.)

ApatheticCosmos commented:
> Code looks much better, only a test failing at this moment.
@gymad I'm trying to figure out why this test is failing:
`1) GoogleSyncTest: Set sync users
 Test  tests/unit/include/GoogleSync/GoogleSyncTest.php:testSetSyncUsers
Failed asserting that 0 is equal to 2 or is greater than 2.
#1  /home/travis/build/salesagility/SuiteCRM/tests/unit/include/GoogleSync/GoogleSyncTest.php:879
#2  GoogleSyncTest->testSetSyncUsers`

It passes when I run it locally, and I've made sure that the users it's counting are the ones added by the test. Yet when Travis runs it, it fails. It sees zero users. Is there something I don't understand about the Travis environment?

gymad commented:
@ApatheticCosmos:
As I checked your code, setSyncUsers() method returns a 0 if one of these are true:
- there is no results of query in travis instance: `"SELECT id FROM users WHERE deleted = '0'"`
Possible solution: make sure you have user before running the test and create it if necessary.

OR:

- the following `if` condition getting `false` when evaluated:
```
if (!empty($user->getPreference('GoogleApiToken', 'GoogleSync')) &&
                    json_decode(base64_decode($user->getPreference('GoogleApiToken', 'GoogleSync'))) &&
                    $user->getPreference('syncGCal', 'GoogleSync') == '1')
```
Possible solution: make sure the following before you run your test:
- GoogleApiToken user preference is not empty,
- GoogleApiToken user preference is base64 encoded JSON string,
- syncGCal user preference is `1`,
(If I were you I adding an exception via specifiend error codes and/or log note if any of these is wrong to see if it is the case)

OR:
- current user ID is not in protected $this->users array. If it is the case please check the log if it contains a warning note added at GoogleSync.php:207 (You can change the log level in travis, see in `.travis.yml`: at line `- echo "\$sugar_config['logger']['level']='error';" >> config_override.php` - If you change the log level please revert it after you fixed this.)

Note: may you already set some necessary environments in your local instance but in travis environments, the tests running on a fresh installed suitecrm. So if you reinstall your CRM locally maybe helps to figurate out what is the issue.

Also in php5.6: As I see the test is failing because: "Argument 1 passed to GoogleSyncBase::getSuiteCRMCalendar() must be an instance of Google_Service_Calendar_CalendarList, null given"
It called at:
```
#1  /home/travis/build/salesagility/SuiteCRM/include/GoogleSync/GoogleSyncBase.php:371
#2  GoogleSyncBase->getSuiteCRMCalendar
#3  /home/travis/build/salesagility/SuiteCRM/tests/unit/include/GoogleSync/GoogleSyncTest.php:235
#4  GoogleSyncTest->testGetSuiteCRMCalendar
```
which means your reflection class has this method, but the argument is passed as `NULL`, however it accepts only an instance of `Google_Service_Calendar_CalendarList`. 
(FYI: only problem with the reflection method invokes as same like the dynamic function calls.. You never able to see if incorrectly using a dependency - fortunately there is a typehint in this case.)

samus-aran commented:
@ApatheticCosmos You are doing a fantastic job! Thanks again for all your hard work and I know Travis can be a pain but these are fantastic steps to growing the feature for future release :+1:

merckaa commented:
I'm having a lot of trouble.
There's something wrong with dates.
When syncinc no valid date in Gcal.
When syncing back no valid date in Meetings. Having problems in convert date in suitecrm logs.

samus-aran commented:
> I'm having a lot of trouble.
> There's something wrong with dates.
> When syncinc no valid date in Gcal.
> When syncing back no valid date in Meetings. Having problems in convert date in suitecrm logs.

Hey @merckaa What locale are you using in both the crm and google? Perhaps we can find a pattern.

merckaa commented:
> > I'm having a lot of trouble.
> > There's something wrong with dates.
> > When syncinc no valid date in Gcal.
> > When syncing back no valid date in Meetings. Having problems in convert date in suitecrm logs.
> 
> Hey @merckaa What locale are you using in both the crm and google? Perhaps we can find a pattern.

Well, actually I found a way to fix Google -> SuiteCRM sync.
I have changed a piece of line to match the date format wanted in the conversion in my current host.

ORIGINAL:
         $event_local->date_start = date("m/d/Y h:i:sa", $starttime);
         $event_local->date_end = date("m/d/Y h:i:sa", $endtime);
CHANGED TO:
        $event_local->date_start = date("d/m/Y H:i", $starttime);
        $event_local->date_end = date("d/m/Y H:i", $endtime);

So a 1 way is doing fine. Now I'm having issues with Sync SuiteCRM to Google.
It appears to not sync if you already have used another Google Account synced before, there should be a way to clean the db from previous syncs.

Tue Dec 18 18:16:05 2018 [3][1][ERROR] ../intranet/include/GoogleSync/GoogleSync.php:316 GoogleSync::syncAllUsers - doSync() Exception: Something went wrong with the requested action
Tue Dec 18 18:16:05 2018 [3][1][ERROR] ../intranet/include/GoogleSync/GoogleSync.php:319 GoogleSync::syncAllUsers - Something went wrong syncing for user id: 1

merckaa commented:
Oh and I have to tell something.
The 404 error seems only to happen when using a subdomain.domain.com
Tested it many times, no luck using subdomain.domain.com but www.domain.com/folder/ seens to sync ok. Not sure why.

ApatheticCosmos commented:
@merckaa I'll need to fix the time format thing. It looks like line 775 is going to need to pull the datetime as a fixed format/locale instead of it's default. I wasn't aware that it did that. I thought it was fixed. Mea culpa. I'll try to find a fix.

My test url has both a subdomain AND a subfolder... so I'm guessing that there is a forward slash missing when it's a domain only. I'll poke around.

merckaa commented:
> @merckaa I'll need to fix the time format thing. It looks like line 775 is going to need to pull the datetime as a fixed format/locale instead of it's default. I wasn't aware that it did that. I thought it was fixed. Mea culpa. I'll try to find a fix.
> 
> My test url has both a subdomain AND a subfolder... so I'm guessing that there is a forward slash missing when it's a domain only. I'll poke around.

Well, about the subdomain/domaing thing, i tried several times and it seens to be when google redirects back and add the SCOPE parameter, it has as url with is not ok with htacess. it gives me a forbidden error.
I just remove que scope parameter in the url and it saves the AuthToken through the code parameter.

About the locale thing, i found a way to bypass the bug with Google -> SuiteCRM sync, but SuiteCRM -> Google sync, uses a DATE_ATOM parameter in line 849 and 854 but changing it breaks the DELETE function.

ApatheticCosmos commented:
@merckaa
The bypass will only work for your locale, as it is the format for dates that's breaking it. However it breaks it for my locale. I need to find a locale agnostic format and use that. Perferably a timezone agnostic one as well. I have to see what the meeting bean will accept.

ApatheticCosmos commented:
@merckaa Try changing those two lines to:
        $event_local->date_start = gmdate("Y-m-d H:i:s", $starttime);
        $event_local->date_end = gmdate("Y-m-d H:i:s", $endtime);

And let me know if that works.

merckaa commented:
> @merckaa Try changing those two lines to:
> $event_local->date_start = gmdate("Y-m-d H:i:s", $starttime);
> $event_local->date_end = gmdate("Y-m-d H:i:s", $endtime);
> 
> And let me know if that works.

Great. It worked in Google -> SuiteCRM sync.
For SuiteCRM -> Google I found a solution for my locale:
Add str_replace like this:
$startDateTime->setDateTime(date(DATE_ATOM, strtotime(str_replace('/', '-',$event_local->date_start))));
Now the sync is OK for my locale, but you may want to add some variations, using if statement maybe.

ApatheticCosmos commented:
Ok. I'll need to figure out what's going on there. startDateTime is an object, not a string. It should be able to accept the timestamp without issue no matter the locale. Weird....

ApatheticCosmos commented:
Btw... what Locale and Timezone are you using, for the record. :)

merckaa commented:
> Btw... what Locale and Timezone are you using, for the record. :)

pt_br | American/Belem

merckaa commented:
Well, I have another problem right now.
The sync is working only for first user, me, admin.
I tested it in others users, it won't work at all. It appears to create the sync job but won't do it.

merckaa commented:
> Well, I have another problem right now.
> The sync is working only for first user, me, admin.
> I tested it in others users, it won't work at all. It appears to create the sync job but won't do it.

I think it's something before setUsersGoogleCalendar() since the SUITECRM calendar is not even created in the second user.

ApatheticCosmos commented:
RC2 is supposed to be tomorrow, and I'd like to get these issues patched up. However, my day job is keeping me very busy today. I'll do my best. :(

merckaa commented:
> RC2 is supposed to be tomorrow, and I'd like to get these issues patched up. However, my day job is keeping me very busy today. I'll do my best. :(

Maybe you can give me just a heads up for adjusting the other users sync.

ApatheticCosmos commented:
@merckaa I can't think of anything that would cause that issue. The users are all set up the same. I'll do some digging once I fix the other issues.

ApatheticCosmos commented:
Still working on updateGoogleCalendarEvent

merckaa commented:
I may have found the problem. the function getusermeetings, user userId variable to retrive the meetings. But user id always return the first id, not the one parsed in the foreach and dosync functions.
Now i'm trying to bypass this.

    protected function getUserMeetings()
    {
        // Validate the workingUser id
        $userId = $this->db->quote($this->workingUser->id);

$userid returns the first id not the foreach id used in syncallusers.

ApatheticCosmos commented:
@merckaa I realize that the getUserMeetings method needs to be modified. It should take the ID as a parameter, not from workinguser. That's leftover cruft. I'm going to clean it up.

merckaa commented:
> @merckaa Please test with the latest commit. See if that clears up all the locale issues.

Ok. I'll try this.

merckaa commented:
> @merckaa Please test with the latest commit. See if that clears up all the locale issues.

@ApatheticCosmos It worked. Thanks for the effort.

ApatheticCosmos commented:
@merckaa Yay!
Still working on the getUserMeetings method... and fixing the tests for these changes.

merckaa commented:
> @merckaa Yay!
> Still working on the getUserMeetings method... and fixing the tests for these changes.

I thing this won't be easy.

merckaa commented:
> @merckaa Yay!
> Still working on the getUserMeetings method... and fixing the tests for these changes.

Found a way to return only the specific id meeting, but when updating it still sync in the first id, it does not rotate.

ApatheticCosmos commented:
@merckaa I think I'm hitting a caching issue when retrieving the user bean, but I'm not sure yet. I'm trying to sort out the tests. If I run the full unit test suite, I get an error in testSetSyncUsers. But if I only run the GoogleSyncTest file, it passes. It's very strange.

merckaa commented:
Well
I have done this:

        $meetings = $this->getUserMeetings($id);

then:

 protected function getUserMeetings($id)
    {
        // Validate the workingUser id
        $userId = $id;

and at last rolledback this function to a older version:

   protected function getClient($id)
    {
        // Retrieve user bean
        if (!isset($this->workingUser)) {
            $this->workingUser = new User();
        }

        $this->workingUser->retrieve($id);

        // Retrieve Access Token JSON from user preference
        $accessToken = json_decode(base64_decode($this->workingUser->getPreference('GoogleApiToken', 'GoogleSync')), true);
        if (!array_key_exists('access_token', $accessToken)) {
        // The Token is invalid JSON or missing
            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Invalid or Missing AuthToken');
            return false;
        }
        // The refresh token is only provided once, on first authentication. It must be added afterwards.
        if (!array_key_exists('refresh_token', $accessToken)) {
            $accessToken['refresh_token'] = base64_decode($this->workingUser->getPreference('GoogleApiRefreshToken', 'GoogleSync'));
        }

        // New Google Client
        $client = new Google_Client();
        $client->setApplicationName('SuiteCRM');
        $client->setScopes(Google_Service_Calendar::CALENDAR);
        $client->setAccessType('offline');
        $client->setAuthConfig($this->authJson);
        $client->setAccessToken($accessToken);

        // Refresh the token if needed
        if ($client->isAccessTokenExpired()) {
            $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Refreshing Access Token');
            $refreshToken = $client->getRefreshToken();
            $client->fetchAccessTokenWithRefreshToken($refreshToken);
            // Save new token to user preference
            $this->workingUser->setPreference('GoogleApiToken', base64_encode(json_encode($client->getAccessToken())), 'GoogleSync');
            $this->workingUser->savePreferencesToDB();
        }
        return $client;
    }

Everything seems to work right now.

ApatheticCosmos commented:
@gymad @samus-aran

I've been really trying to get this done for RC2 tomorrow, but I've hit a snag with tests. If you remember during the pull party every time I'd run the tests locally, they'd work, but then fail when travis ran them.
The specific method that's failing is GoogleSync::setSyncUsers(). The line:

`$user = BeanFactory::getBean('Users', $row['id']);`

Is returning a null bean, even though the id is populated.

I think I've tracked it down. Since the switch to phpunit, running ALL the unit tests causes this failure locally, but running just the GoogleSyncTest class would not.
I've traced it down to SugarBean::populateFromRow.
When running all the tests, field_defs only has a couple of items. However when running just GoogleSyncTest it has the full list you'd expect.

I'm at a loss on this. I have no idea why this is happening, and need some expert help.

merckaa commented:
@ApatheticCosmos 
I don't know if this is a feature that already exists but when you create a meeting, and then sync in one user and then if you need to change the user, it breaks the sync. It is not possible to change the user after the meeting was created and synced. Don't know if this is already a feature, but if it is, this is broken too.
If it's not a feature, it's something to think about for the future.

ApatheticCosmos commented:
@merckaa That's not something I've considered. I have a thought in my head as to how to handle it, but It'll  have to wait until I get this PR sorted out. Thanks for the input though!

ApatheticCosmos commented:
@merckaa Also I've noticed that since I merged upstream changes, I can't toggle the 'Enable Calendar Sync' checkbox. I'll have to figure that out too.

ApatheticCosmos commented:
@merckaa
Try it all now. The latest commit should let multiple users sync, and the 'Enable Calendar Sync' checkbox should be saving.
Something I learned as a heads up: The id of demo users don't pass SuiteValidator::isValidId() :P
They will fail if you use them for testing this... unless you change that method in lib/Utility/SuiteValidator.php to just return true for testing.

ApatheticCosmos commented:
@merckaa / Anyone else
Join me at https://gitter.im/SuiteCRM_AC/community for live chat about this pull request.

merckaa commented:
> @merckaa / Anyone else
> Join me at https://gitter.im/SuiteCRM_AC/community for live chat about this pull request.

So sorry, very busy at work today. =(

ApatheticCosmos commented:
I really, really wish I could duplicate the Travis environment locally.

ApatheticCosmos commented:
Oh come on Travis! Really?

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-            $this->gClient = $gClient_local;
-            return true;
+        if (!$meeting) {
+            // TODO: should it be a user message?
```
@gymad
I don't think this needs to be logged at all. It would be completely normal for this function to only get one event or the other. If both are missing, it sets it to "UNNAMED RECORD". No error.

ApatheticCosmos replied:
Fixed in temp branch

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'ERROR:Missing Variables');
+        $meetings = $this->getUserMeetings($id);
+        if (empty($meetings)) {
+            // TODO: It should be an exception
```
@gymad
Oh man... I'm glad you had me look at this.
There is a possibility that $meetings could be empty, if the user doesn't have any SuiteCRM meetings at all.
This wouldn't be an error, and the process should continue, possibly pulling meetings from Google.
I'll have to fix this.

ApatheticCosmos replied:
Fixed in temp branch

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
+        try {
+            $this->initUserService($id);
+        } catch (Exception $e) {
+            // TODO: handle the exception instead just loggin it, it will be logged if unhandled by suitecrm core.
```
@gymad 
All the Exceptions caught/thrown here are handled in syncAllUsers(). I suppose this log output is redundant.

ApatheticCosmos replied:
removed redundant logging.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-                $reminderInvitee->related_invitee_module_id = $this->workingUser->id;
-                $reminderInvitee->save(false);
+            if (!$actionResult) {
+                // TODO: confusing info log: $actionResult evaluated as string but it also false if converted to boolean. There is an empty string in log as result. - TODO: Inform the caller about this problem (exception?)
```
@gymad 
Yep. Very confusing. I think this is left over cruft. doAction() was changed significantly. It's never going to return a string, only true or throw exception.

ApatheticCosmos replied:
Removed if statement

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $results_g = $this->gService->events->listEvents($this->calendarId, $optParams);
+
+        // We only want the events, not the leading cruft
+        $results = $results_g->getItems(); // TODO: getItems() returns a Google_Service_Calendar_Event, but this code expect an array in the next line? Something missing here?
```
@gymad
listEvents() above returns Google_Service_Calendar_Events
getItems() returns an array of Google_Service_Calendar_Event objects.
Google made the names a bit confusing.

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'ERROR:Missing Variables');
+        // Now, we look at the Google Calendar
+        $googleEvents = $this->getUserGoogleEvents();
+        if (!isset($googleEvents)) { // TODO: if condition never be true, as $googleEvents is definitely set just one line before.
```
@gymad 
More cruft. this was used to catch if getUserGoogleEvents() returned null. I'll clean it up.

ApatheticCosmos replied:
Done

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-                $reminder_remote->setMinutes($reminder_local->timer_popup / 60);
-                $reminders_array[] = $reminder_remote;
+            if (!$actionResult) {
+                // TODO: Inform the caller about this problem (exception?)
```
@gymad 
Cruft. doAction() will throw an exception on error. This will never get triggered.

ApatheticCosmos replied:
fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -46,1088 +53,249 @@
-        if (!$this->setClient($id)) {
-            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Unable to setup Google Client');
+        if (empty($meeting) && empty($event)) {
+            // TODO: It should be an exception
```
@gymad 
Agreed. I'll fix it.

ApatheticCosmos replied:
done

ApatheticCosmos commented on include/GoogleSync/GoogleSync.php:
```diff
@@ -1143,25 +311,43 @@ public function syncAllUsers()
+        try {
+            $ret = $this->setSyncUsers();
+        } catch (Exception $e) {
+            // TODO: handle the exception instead just loggin it, it will be logged if unhandled by suitecrm core.
```
@gymad 
I'm not sure what do to here. I don't want the syncAllUsers() to stop here if a single user has a configuration issue that throws an exception. It should move on to the next user. After it's all done, if there were any errors, it will return false to the cron runner so it's marked as unsuccessful.
Should there be some kind of user notification given if it fails? Maybe send the user a email, or some kind of popup on their screen. I don't know. At this point I'm not sure increasing the scope of the project is advisable. I'm already crammed for time. I really need to get this PR pulled so the fixes are in the next RC.

ApatheticCosmos replied:
I'll just let SuiteCRM log the exception for now.

ApatheticCosmos replied:
Err... I can't do that. It stops the foreach loop from progressing to the next loop. I'm catching it here, and logging it. It's all I can do with it at this point. I'm logging a simple 'exception caught' message to fatal, and a stack trace to error.

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        // This sets the log level to a variable that can be set on the command line while running cron.php on the server. It's for debugging only.
+        // EXAMPLE: $ GSYNC_LOGLEVEL=debug php cron.php
+        $this->logger = LoggerManager::getLogger();
+        if (isset($_SERVER['GSYNC_LOGLEVEL'])) { // TODO: is it possible to using $sugar_config instead superglobals? (a global usage is one step closer to clean code than using a superglobal)
```
@gymad 
How about I add a line to config_override.php for this. That's almost as easy for the user. It's only ever used for debugging.

ApatheticCosmos replied:
Done

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $this->logger = LoggerManager::getLogger();
+        if (isset($_SERVER['GSYNC_LOGLEVEL'])) { // TODO: is it possible to using $sugar_config instead superglobals? (a global usage is one step closer to clean code than using a superglobal)
+            $this->oldLogLevel = $this->logger->getLogLevel();
+            $this->logger->setLevel($_SERVER['GSYNC_LOGLEVEL']); // TODO: it changes the log level but as it is a global, it will affect the other parts of core until this class __destructor is called! (and it's bad)
```
@gymad 
Yes. That is by design. It's only used for debugging. With this I can have a user set the debug level only for the sync process. It will log at that level for all child functions. This way they only see the debug log, for instance, of the sync job.
I don't think that this sets the log level globally for the whole system. It will only set it for children of this function. Am I incorrect? I've never seen anything other than sync related logging when using this. Nothing else.

gymad replied:
Two point is not clear for me here:
If it's needed for debugging only why not just simply check for e.g the isDeveloperMode() function instead introducing a new superglobal. Or for e.g you also can check a $sugar_config variable which is specially for this and add it to your config_override.php instead introduce new superglobals...
Second problem what I can see is you restore the log level only in destructor. It means everything (but really everything not just your new code and functions) will using the log level as defined in `$_SERVER['GSYNC_LOGLEVEL']` after the constructor called until this destructor set it back. 
Maybe more clear if I take this into an example: 

Here is a very simple php file (only for example..)
```
<?php
hereDoingThings();  // <-- using original log level. .. OK :)
?>
```

Now I want to add your code into this because at this php I want to Sync my google accouns but I don't want to brake anything in original behaviour of this but I will do. (In other words, where your destructor called in the following example php file after this modification?)

```
<?php
 // it will change the log level:
$sync = new GoogleSync(..);
$sync->doingThingsWithNewLogLevel(...); // using debug level .. OK

// here everything else using your specified log level but shouldn't.
// It because the LoggerManager::$_instance is static pointer to a logger instance
// and LoggerManager::getLogger() is a singleton factory (antipattern unfortunatelly)

hereDoingThings();  // <-- using debug level, however I did not change it .. BAD :(

$sync->doingSomeOtherThings(); // still debug level.. OK

// php file ended, GoogleSyncBase::__destruct called and restoring log level only at this point.
?>
```

As you see, the logger badly implemented as a static singleton. I know it is a wrong desing but unfortunatelly it's already added in this way and it means we should be carefull.

ApatheticCosmos replied:
@gymad Ok. I understand now. The Logger is not an independent object, as I thought it was. The whole LoggerManager class is forcing this. Ugh...
Even __construct is a private method. :man_facepalming:

How about if I explicitly destroy the class after running the sync by adding '$googleSync = null' in the syncGoogleCalendar() function modules/Schedulers/_AddJobsHere.php ?
Would that be enough to keep this logging functionality? I don't know any other way around this. I'll have to remove the custom logging.

PS. I had already changed from using a superglobal to an entry in config_override.php.

ApatheticCosmos replied:
I've removed the logging code.

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $accessToken = json_decode(base64_decode($this->workingUser->getPreference('GoogleApiToken', 'GoogleSync')), true);
+        if (!array_key_exists('access_token', $accessToken)) {
+            // The Token is invalid JSON or missing
+            // TODO: It should be an exception (do not leave same 'false' return value for differenc failure reason, Inform the caller about this problem)
```
@gymad 
I'll look into this. The return is passed to setClient, and then initUserService, which then throws an exception if it failed.
I'll try to figure out a cleaner way to handle it.

ApatheticCosmos replied:
Done. I'm going to need to do some extensive real world tests to make sure this didn't break anything.

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        try {
+            $client = $this->getGoogleClient($accessToken);
+        } catch (Exception $e) {
+            // TODO: handle the exception instead just loggin it, it will be logged if unhandled by suitecrm core.
```
@gymad 
I want to make sure that the sync process can continue for other users if the current one fails. If I leave this uncaught, will it stop the scheduled job when thrown?

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        }
+
+        if (!$this->setClient($id)) {
+            // TODO: It shouldn't be in the log as it is an exception, it will be logged if unhandled by suitecrm core.
```
@gymad 
Yeah this is redundant. I'll remove the logging

ApatheticCosmos replied:
Done

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+
+        // if the SuiteCRM calendar doesn't exist... Create it!
+        if (!$this->isCalendarExists()) {
+            // TODO: It should be an exception?
```
@gymad 
I don't think it should be an exception... if it doesn't exist, it gets created. Now down at "// Final Check" we may want to throw something. I'll look into it.

ApatheticCosmos replied:
Done

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+                break;
+            }
+        }
+        if (empty($return)) { // TODO: this empty-check is not neccessary, return the getId() results as it is.
```
@gymad
I'll have to refresh my memory on this. It may be that getId() returns an empty string or array, and I need a null.

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+    {
+        // Make sure we have a calendar id
+        if (!isset($this->calendarId)) {
+            // TODO: I think a warning enough into log instead fatal (but it should be handled by caller function)
```
@gymad 
Heck yeah... that's a bit excessive. I'll fix it.

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+    {
+        // Make sure we have a Google Calendar Service instance
+        if (!isset($this->gService)) {
+            // TODO: I think a warning enough into log instead fatal (but it should be handled by caller function)
```
@gymad
Same as below. Will fix.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+
+        // This checks to make sure we only get one result. If we get more than one, something is inconsistant in the DB
+        if ($result->num_rows > 1) {
+            // TODO: It shouldn't be in the log as it is an exception, it will be logged if unhandled by suitecrm core.
```
@gymad
I'll clean this up... but I want to make it clear that this is a 'scream from the hilltops' error. More then one row returned means something has gone horribly wrong with the data in the DB. Corruption.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+            $this->logger->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'More than one meeting matches Google Id: ' . $eventIdQuoted);
+            throw new GoogleSyncException('More than one meeting matches Google Id!', GoogleSyncException::AMBIGUOUS_MEETING_ID);
+        } elseif ($result->num_rows == 0) {
+            // TODO: Confusing return value: Do not use similar return value by different reason of failure. see belove the $meeting->retrive
```
@gymad
This is not a failure condition. It means the meeting exists on Google, but not here. It returns null and pushPullSkip then knows to pull it.

ApatheticCosmos replied:
Cleaned it up a bit

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+    {
+        // make sure we have a client set
+        if (!isset($this->gClient)) {
+            // TODO: Confusing return value: It should be an exception, do not use similar return value (false) by different reaon os failure (see belove)
```
@gymad 
I think you're right. It should throw an exception. Will fix.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+         * So we check to make sure it has an ID to determine Success/Failure.
+         */
+        if (!isset($return->id)) {
+            // TODO: Confusing return value: It should be an exception, do not use similar return value (false) by different reaon os failure (see belove)
```
@gymad 
I think I could throw and then catch it in doAction(). Then I could replace the exception message there with the message/code from this.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        }
+
+        if (empty($event)) {
+            // TODO: Confusing return value: It should be an exception, do not use similar return value (false) by different reaon os failure (see belove: setLastSync() also could returns a 'false')
```
@gymad 
Same here. Throw and catch in doAction().

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        
+        // Make sure the calendar service is set up
+        if (!$this->isServiceExists()) {
+            // TODO: Confusing return value: It should be an exception, do not use similar return value (false) by different reaon os failure (see belove)
```
@gymad 
I'm seeing a pattern. Throw here and catch in doAction().

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+    protected function clearPopups($event_id)
+    {
+        if (!isset($event_id) || empty($event_id)) {
+            // TODO: Confusing return value: It should be an exception, do not use similar return value (false) by different reaon os failure (see belove)
```
@gymad 
Yep. I'll fix.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $event_local->description = (string) $event_remote->getDescription();
+        $event_local->location = (string) $event_remote->getLocation();
+
+        // TODO: dont leave todo comment in code:
```
@gymad 
I'll change it to something humorous. Would a 'FUTURE' tag work?

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $event_id = $event_local->id;
+        $res = $this->clearPopups($event_id);
+        if (empty($res)) {
+            // TODO: Inform the caller about this problem (exception or return value?)
```
@gymad 
I'll need to look at what clearpopups returns, and fix it.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+    protected function updateGoogleCalendarEvent(Meeting $event_local, Google_Service_Calendar_Event $event_remote)
+    {
+        if ((!isset($event_local) || empty($event_local)) || (!isset($event_remote) || empty($event_remote))) {
+            // TODO: It should be an exception
```
@gymad 
yes it should. I'll fix.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+            $this->logger->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Timezone set to \'' . $this->timezone . '\'');
+            return true;
+        }
+        // TODO: I think a warning enough into log instead fatal (but it should be handled by caller function)
```
@gymad 
I'll look into this.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+            //$this->syncedList[] = $event_local->id;
+            return $event_local->id;
+        }
+        // TODO: I think a warning enough into log instead fatal (but it should be handled by caller function)
```
@gymad 
If it gets here, something bad happened. I'll find an appropriate way to handle it.

ApatheticCosmos replied:
Fixed

ApatheticCosmos commented on include/GoogleSync/GoogleSyncBase.php:
```diff
@@ -0,0 +1,1049 @@
+        $authJson_local = json_decode(base64_decode($sugarConfig['google_auth_json']), true);
+        if (!$authJson_local) {
+            // The authconfig json string is invalid json
+            // TODO: It should be an exception (do not leave same 'false' return value for differenc failure reason, Inform the caller about this problem)
```
Fixed

Abuelodelanada commented:
> @horus68 I'm not sure what to do at this point. The language file has other examples of both escaping a single quote/apostrophe (like LBL_USE_REAL_NAMES_DESC), and using HTML codes for special symbols (like LBL_SETTINGS_URL_DESC).
> I don't know of any other solutions to the problem. frowning_face
> How should this be handled? I could spell it out as 'do not' but that can't be a solution all the time. There's too many contractions commonly used in english to ban them all.
> 
> There's one possible solution I just thought of... could we use a unicode [’] instead? It seems to work.
> It looks like this in the code:
> ![image](https://user-images.githubusercontent.com/16560296/49162784-a4c5b900-f2f9-11e8-91af-4451ac46add7.png)
> 
> And like this when rendered by chrome:
> ![image](https://user-images.githubusercontent.com/16560296/49162761-937cac80-f2f9-11e8-9590-2f152745c3f3.png)
> 
> What are your thoughts on this?

Hi @ApatheticCosmos 

What about this translation:

`'LBL_GOOGLE_API_TOKEN_HELP' => "Use this to authorize SuiteCRM to access your Google Calendar. This is required for synchronization. Don't forget to check the 'Enable Calendar Sync' checkbox when done.",`

horus68 commented:
I can't make a decision here and i'm not a coder, I can only say when something work for Crowdin or not and after testing.
But we can't make a solution that will not allow translators to use the words as their language uses (French language has too many problems with translations until now - not just for this string - and there are so many languages and alphabets around here!).
Until now there isn't a pattern for all files on single/double quotes, not even in language keys format.
So its up to SA to find a solution that allows non escaped signs in language values and its translations.
Calling @Dillon-Brown to the conversation!

ApatheticCosmos commented:
@horus68 @Dillon-Brown 
The root question I have then is this: Can we use Unicode characters in the label strings?
The answer to this question will determine my next commit :)

Dillon-Brown commented:
@ApatheticCosmos Hmm, I'd probably prefer just changing the label for now. We do need to come up with a better solution for the language strings but it's not something that can be done in this PR.

ApatheticCosmos commented:
@Dillon-Brown
Ok, I'm going to escape the single_quote/apostrophe for now. It's the simplest thing to understand at this point.
Hopefully we come up with a better solution in the future that doesn't require any escaping, and we can apply it globally then.

gymad commented:
@ApatheticCosmos Can you please revert the additional TODO messages please, I can not find them so I can not fix them. Also next time please do not switch off the state checker in travis because impossible to see if some test left some wrong data. Also please do not use reflection class even in tests because xdebug can not follow the stack trace if it has no any. Also please do not never leave a catch as empty in a try..catch statement.
I close this PR as it outdated. See these changes in a fixed PR: https://github.com/salesagility/SuiteCRM/pull/6696

ApatheticCosmos commented:
@gymad 
> @ApatheticCosmos Can you please revert the additional TODO messages please, I can not find them so I can not fix them. 
I don't know what you mean by reverting additional TODO messages. I removed the ones you added as I fixed the todo's.

>Also next time please do not switch off the state checker in travis because impossible to see if some test left some wrong data.
I don't recall doing this?

>Also please do not use reflection class even in tests because xdebug can not follow the stack trace if it has no any.
Using reflection is the only way to test non-public methods. Is there another way? I thought you said you liked how I used the reflection class?

>Also please do not never leave a catch as empty in a try..catch statement.
Ok

> I close this PR as it outdated. See these changes in a fixed PR: #6696
Is it ready to be merged? If I'm to do any more work on it I'm going to need to be able to push changes to the PR.

pgorod commented:
@ApatheticCosmos does this bug look like something that is already fixed in the newest code?

https://suitecrm.com/suitecrm/forum/suitecrm-beta/21471-i-can-not-set-enable-calendar-sync

ApatheticCosmos commented:
@pgorod yes. I remember that issue. 
It's fixed in this pr.

QuentinDoe commented:
Hello , i dont know if it's resolved in new version , i make upgrade 7.11.3 and i can not set enable calendar sync
